sdk/inc/azure/core/az_mqtt5_rpc_client.h

```diff
@@ -0,0 +1,334 @@
+   */
+  AZ_EVENT_RPC_CLIENT_READY_IND = _az_MAKE_EVENT(_az_FACILITY_CORE_MQTT5, 24),
+  /**
+   * @brief Event representing the application requesting to send a command.
+   */
+  AZ_EVENT_RPC_CLIENT_INVOKE_REQ = _az_MAKE_EVENT(_az_FACILITY_CORE_MQTT5, 25),
+  /**
+   * @brief Event representing the RPC client receiving a command response and sending it to the
+   * application
+   */
+  AZ_EVENT_RPC_CLIENT_RSP = _az_MAKE_EVENT(_az_FACILITY_CORE_MQTT5, 26),
+  /**
+   * @brief Event representing the application requesting the RPC client to unsubscribe from the
+   * response topic
+   */
+  AZ_EVENT_RPC_CLIENT_UNSUB_REQ = _az_MAKE_EVENT(_az_FACILITY_CORE_MQTT5, 27)
+};
+
+struct az_mqtt5_rpc_client_hfsm
+{
+  struct
+  {
+    /**
+     * @brief RPC Client policy for the MQTT5 RPC Client.
+     *
+     */
+    _az_hfsm rpc_client_policy;
+
+    /**
+     * @brief The subclient used by the MQTT5 RPC Client.
+     */
+    _az_event_client subclient;
+
+    /**
+     * @brief The MQTT5 connection linked to the MQTT5 RPC Client.
+     */
+    az_mqtt5_connection* connection;
+
+    /**
+     * @brief the message id of the pending subscribe for the response topic
+     */
+    int32_t pending_subscription_id;
+
+    /**
+     * @brief timer used for the subscribe of the response topic
+     */
+    _az_event_pipeline_timer rpc_client_timer;
+
+    /**
+     * @brief az_mqtt5_rpc_client associated with this hfsm
+     */
+    az_mqtt5_rpc_client* rpc_client;
+
+    /**
+     * @brief The property bag used by the RPC client policy for sending request messages
+     */
+    az_mqtt5_property_bag property_bag;
+
+  } _internal;
+};
+
+// Event data types
+
+/**
+ * @brief Event data for #AZ_EVENT_RPC_CLIENT_INVOKE_REQ.
+ */
+typedef struct az_mqtt5_rpc_client_invoke_req_event_data
+{
+  /**
+   * @brief The correlation id of the request.
+   */
+  az_span correlation_id;
+
+  /**
+   * @brief The content type of the request.
+   */
+  az_span content_type;
+
+  /**
+   * @brief The payload of the request.
+   */
+  az_span request_payload;
+
+  /**
+   * @brief The message id of the request to correlate with pubacks.
+   */
+  int32_t mid;
+} az_mqtt5_rpc_client_invoke_req_event_data;
+
+/**
+ * @brief Event data for #AZ_EVENT_RPC_CLIENT_RSP.
+ */
+typedef struct az_mqtt5_rpc_client_rsp_event_data
+{
+  /**
+   * @brief The correlation id of the response.
+   */
+  az_span correlation_id;
+
+  /**
+   * @brief The error message of the response.
+   */
+  az_span error_message;
+
+  /**
+   * @brief The status of the response.
+   */
+  az_mqtt5_rpc_status status;
+
+  /**
+   * @brief The content type of the response.
+   */
+  az_span content_type;
+
+  /**
+   * @brief The payload of the response.
+   */
+  az_span response_payload;
+
+  /**
+   * @brief Whether the RPC client failed to parse the response - if the correlation id was able to
+   * be parsed, the request can still be considered completed if desired.
+   */
+  bool parsing_failure;
```

Comment:
I don't really like this: 
1. I'm not sure what it means that the correlation is correct but parsing still failed later: maybe the correlation ID is also wrong (even if it appears correct).
2. Instead of this `bool` I propose two separate event types if partial parsing is really desired: `AZ_EVENT_RPC_CLIENT_RSP` and `AZ_EVENT_RPC_CLIENT_PARSE_ERROR_RSP`.

`AZ_EVENT_RPC_CLIENT_PARSE_ERROR_RSP` could have an identical data structure and we could document that we've performed "best-effort" decoding.

Still, the application code would really need to be aware of this data being at most partial (if not completely corrupted).